### ISSUE TITLE
Can't verify domain: [WARNING: No SPF record exists for this domain]

### ISSUE THREAD

**Issue by shinebayar-g** [atech/postal#623]
Hello postal team.
First of all, thanks for making such great project. Unfortunately I'm having hard time setting this up.
I tried to solve this problem all myself without opening issue, but It seems I can't get it working without help.

I've successfully installed the postal project and everything runs fine. I created my first organization and first server. Went to **Domains** and added my first domain. Followed the instruction as it said, but I can't verify **SPF Record** part.

Here is my `postal.yml` file. I removed unrelated parts.

```YAML
web:
  # The host that the management interface will be available on
  host: postal.mydomain.io
  # The protocol that requests to the management interface should happen on
  protocol: https

dns:
  # Specifies the DNS record that you have configured. Refer to the documentation at
  # https://github.com/atech/postal/wiki/Domains-&-DNS-Configuration for further
  # information about these.
  mx_records:
    - postal.mydomain.io
  smtp_server_hostname: postal.mydomain.io
  spf_include: spf.postal.mydomain.io
  return_path: rp.postal.mydomain.io
  route_domain: routes.postal.mydomain.io
  track_domain: track.postal.mydomain.io
  helo_hostname: # By default, this will be the same as the `smtp_server_hostname`
  dkim_identifier: postal
  domain_verify_prefix: postal-verification
  custom_return_path_prefix: psrp

smtp_server:
  port: 2525
  tls_enabled: false
  tls_certificate_path: # Defaults to config/smtp.cert
  tls_private_key_path: # Defaults to config/smtp.key
  proxy_protocol: false
  log_connect: true
  strip_received_headers: false

smtp_relays:
  -
    hostname:
    port: 2525
    ssl_mode: Auto

smtp:
  # Specify an SMTP server that can be used to send messages from the Postal management
  # system to users. You can configure this to use a Postal mail server once the
  # your installation has been set up.
  host: postal.mydomain.io
  port: 2525
  username: myfirst-stmp # Complete when Postal is running and you can
  password: HvVmXCJeeqeZdyVvGc3bmvb7I # generate the credentials within the interface.
  from_name: Postal
  from_address: noreply@postal.mydomain.io
```

My DNS entries:

```
A | postal | points to MyVpsStaticIP
CNAME | psrp.postal | is an alias of rp.postal.mydomain.io
MX | postal | mail handled by postal.mydomain.io (10)
TXT | postal-cayfef._domainkey.postal | v=DKIM1; t=s; h=sha256; p=MIGfMA0GCSqGSIb3DQEB...
TXT | mydomain.io | v=spf1 include:servers.mcsv.net include:spf.postal.mydomain.io ?all
```

About last record, there was an existing TXT record on my root domain **mydomain.io**. So I added **include:spf.postal.mydomain.io** to existing record as Dns setup page suggested. Since it didn't work I even removed existing record and added new entry for `domain root: @` `v=spf1 a mx include:spf.postal.mydomain.io ~all`

I've tried to send test email, but it seems I cannot send emails until I verify the domain. It says SOFT FAIL.

**Comment by oasis1234** [comment]
Please, don't say that you "followed installation instructions step by step" (referring to my guide https://github.com/atech/postal/issues/614) when you didn't do it at all.
My guide don't ask you to include " dkim_identifier: postal,   domain_verify_prefix: postal-verification,   custom_return_path_prefix: psrp".
My guide gives you **13 DNS records, and you only include 5**. What?! Please go back and read again. Just replace "mail" with "postal", and "example.com" with "mydomain.io".
I assume that you know how to configure DNS records properly. For example, you have created a CNAME for the subdomain "psrp.postal", when it's just "psrp".
You created a MX for the subdomain "postal", but not for the root domain @.
And the list goes on...

**Comment by shinebayar-g** [comment]
I didn't mention your guide at all here. I tried to keep up with your guide on your guide page, it didn't happen. I guess talking about your guide here is not related to main project issues, unless I specifically mention you here. Above setup and results are according to **Dns Setup** page.

About your guide I already tried several times (yes I created all of the DNS records as you said there) but only DKIM step was green and other 3 weren't verified.

If you don't mind to help me setup according to your guide, I will reconfigure/redo all my setup.

PS: Please forgive my English.

**Comment by shinebayar-g** [comment]
Ok, I'm following your guide. @oasis1234 
```
@	A	1.2.3.4          (Is this necessary?, My root domain.io is hosting other website on other server.)
postal	A	1.2.3.4
rp.postal	A	1.2.3.4
track.postal	A	1.2.3.4
psrp	CNAME	rp.postal.mydomain.io
@	MX	10;postal.mydomain.io
routes.postal	MX	10;postal.mydomain.io
rp.postal	MX	10;postal.mydomain.io
@	TXT	v=spf1 a mx include:spf.postal.mydomain.io ~all
postal-a1b2c3._domainkey	TXT	v=DKIM1; t=s; h=sha256; p=MIGfMA0GC...;
rp.postal	TXT	v=spf1 a mx include:spf.postal.mydomain.io ~all
spf.postal	TXT	v=spf1 ip4:1.2.3.4 ~all
_dmarc	TXT	v=DMARC1; p=reject; sp=reject; rua=mailto:postmaster@mydomain.io
```

Hmm, none of 4 domains are green. Am I missing something?

Here is steps I done.

- [x] Install postal dependencies.

- [x] Install, configure and run postal.

- [x] Set these Dns records in Cloudflare.

**Comment by oasis1234** [comment]
Let's focus on just one DNS record.
1. You have in Cloudflare: `Type: TXT, Name: mydomain.io, Value: v=spf1 a mx include:spf.postal.mydomain.io ~all, TTL: Automatic, Status x 
2. You checked in whatsmydns.net that mydomain.io TXT gives that value.
Restart Postal (just in case) `postal restart`
3. Go to the DNS Setup page and click "Check my records".
If you did 1, 2 and 3, it's mathematically impossible that the verification fails. If that happens, post screenshots of everything [you will delete them later]. I need see to believe.

**Comment by shinebayar-g** [comment]
Sorry for late reply. I removed all DNS records related to postal except postal.mydomain.io A record to my server. 

1. Added the TXT record on cloudflare (Here is all TXT records of my domain)
![image](https://user-images.githubusercontent.com/3091558/43440003-44bfb932-94c8-11e8-9b27-535ab26b6f14.png)
2. Checked status on whatsmydns.net (I assume it's working?)
![image](https://user-images.githubusercontent.com/3091558/43440113-a94b4cc2-94c8-11e8-9b15-be9b7f823e05.png)
3. Restarted the postal service. (I'm using instructions from https://github.com/atech/postal/wiki/Starting-on-boot page. `sudo systemctl restart postal` to restart postal app)
![image](https://user-images.githubusercontent.com/3091558/43440241-2a293ea8-94c9-11e8-8ea6-0ba2bc72791d.png)

Check my records result:
![image](https://user-images.githubusercontent.com/3091558/43440327-75452c12-94c9-11e8-80a9-299a0c5d1b06.png)

**Comment by willpower232** [comment]
It looks like you have two SPF records. This will not work. You should combine all the values into one record.

**Comment by willpower232** [comment]
A better record would be `v=spf1 a mx include:servers.mcsv.net include:spf.postal.erxes.io ~all`

**Comment by shinebayar-g** [comment]
Done.
Should I restart postal after any DNS changes?

**Comment by willpower232** [comment]
That shouldn't be necessary.

**Comment by shinebayar-g** [comment]
Well I guess there is something wrong with my setup. It really should be working like this right?
What should be causing this issue?

- Cloudflare settings? ( I even enabled development mode)
- Nginx settings? (All I changed was: `mail.example.com` -> `postal.mydomain.io`)
- Firewall? (I changed SMTP server ports to 2525 and opened them on firewall. I guess It's working as expected)
- Google Cloud could be blocking something somewhere?

**Comment by oasis1234** [comment]
I'm sorry. Your case is difficult and my knowledge of Unix is limited and Ruby almost inexistent.
I spent days trying to use a proxy with Postal (issue 605). Nobody helped me. If you are a "die-hard guy" like me, I recommend you to spend hours analyzing the source code and doing tests. Maybe start from the scratch with a VM system and a test domain (freenom).

If nobody else here helps you, you will need to continue your way alone. Just a final tip for you:
I have tested that having several records its not a problem. Read the source code
https://github.com/atech/postal/blob/master/app/models/domain/dns_checks.rb#L44
It seems to me that there is some problem when Postal code tries to get records, as it gives the error "empty".
Create a file named testdns.rb with this
```
require 'resolv'
resolver = Resolv::DNS.new
result = resolver.getresources('erxes.io', Resolv::DNS::Resource::IN::TXT)
puts result.inspect.gsub(",", "\n")
```
Run ruby testdns.rb an see what outputs.
(My system outputs now: 
```
 #<Resolv::DNS::Resource::IN::TXT:0x00563335e24438 @strings=["v=spf1 a mx include:servers.mcsv.net include:spf.postal.erxes.io ~all"]
 @ttl=300>]
```
```

**Comment by shinebayar-g** [comment]
Well, tomorrow I will create new server and domain for the sake of this application testing.

**Comment by oasis1234** [comment]
Just out of curiosity, what's the output of running the code I posted (testdns.rb) ?

**Comment by shinebayar-g** [comment]
@willpower232  yes, that's why I changed port 25 to 2525. Most server providers doesn't allow 25, 465, 587 ports nowadays.
@oasis1234 I will let you know when I get to work tomorrow.

**Comment by shinebayar-g** [comment]
@oasis1234 
I've created file named `testdns.rb` on my home directory and run the script using 
`ruby testdns.rb`. Is this right?

And output is:

```Shell
mailserver:~$ ruby testdns.rb 
[#<Resolv::DNS::Resource::IN::TXT:0x0000000000d81288 @strings=["v=spf1 a mx include:servers.mcsv.net include:spf.postal.erxes.io ~all"]
 @ttl=258>
 #<Resolv::DNS::Resource::IN::TXT:0x0000000000d80ab8 @strings=["google-site-verification=-tepODKHoWJkXll9WN-BXI6HtFm3_u0-3rNXYWWfPyc"]
 @ttl=258>]
```

It's still not working, I guess I'm gonna start building test environment now.

**Comment by shinebayar-g** [comment]
Well shit I either tried https://github.com/atech/postal/wiki/Domains-&-DNS-Configuration, https://github.com/atech/postal/issues/614 and Dns Setup page instructions. None of them worked for me on clean DigitalOcean Ubuntu 16.04 vps and new clean domain.

Unless someone teamviewer me and go through whole process I think I'm done here.

**Comment by willpower232** [comment]
Oh hang on, you're adding `postal.erxes.io` but you've put the SPF record on `erxes.io`

**Comment by oasis1234** [comment]
> you've put the SPF record on erxes.io

Yes, that's what Postal is asking for:

> You need to add a TXT record at the apex/root of your domain (@) with the following content

You can review the source code (dns_checks.rb#L44). The only solution for shinebayar was to debug the source code.

**Comment by willpower232** [comment]
@oasis1234 You're right and the code is probably wrong since the policy on the root of the domain shouldn't always apply to the subdomains (http://www.openspf.org/FAQ/The_demon_question)

@shinebayar-g can you add other top level domains to Postal? i.e. `erxes.io` should be able to be added without changing your existing settings.

If you can't, it might be worth switching how Postal looks up DNS with the `use_local_ns_for_domains` setting in your config file (see the default at https://github.com/atech/postal/blob/master/config/postal.defaults.yml#L15)

**Comment by oasis1234** [comment]
**TL;DR Postal checks the DNS records based on the label "Name" entered when adding a domain, ignoring the postal.yml file.**

@shinebayar-g maybe you're not going to read this, or you're no longer interested, but I'm a stubborn guy that never gives up.
Today I was installing Postal in a new server & domain, let's call it mydom.com
I clicked "Add new domain", then an input box with label "Name" appears, so I simply entered "mydom", knowing that everything was well configured in postal.yml and the DNS table. Then I went to the DNS Setup page, clicked 'Check my records are correct' and... KaBOOM! Postal complained that all DNS records were wrong. I rechecked everything, and noticed that the error said "No txt records were returned for postal-abc123._domainkey.mydom" Notice that it said "mydom", not "mydom.com". So I deleted the domain and created a new with mydom.com
@shinebayar-g You entered postal.erxes.io as your domain, you should have only entered erxes.io. Sorry for the late reply.
Final thoughts: "If your neighbor has a problem, be prepared, maybe you're next".

**Comment by shinebayar-g** [comment]
Alright, you convinced my man. I'm gonna create new server and domain right now. I will write down all of my steps I'm doing. Hope it's gonna work this time.

**Comment by oasis1234** [comment]
You're welcome!

**Comment by gitaayush** [comment]
@oasis1234  I am having similar issues with the DNS record. I have maintained all the record required by postal as well as mentioned on the postal for dummies post. 

Also testdns.rb outputs following result

ruby testdns.rb
[#<Resolv::DNS::Resource::IN::TXT:0x0000559c47955400 @strings=["v=spf1 a mx include:spf.relay.cloudhimalaya.com.np ~all"]
 @ttl=3036>]

I am missing something

**Comment by willpower232** [comment]
@gitaayush please feel free to open a new issue rather than replying to a closed one

**Comment by abewartech** [comment]
I Have same problem to

ciptapernika.id	3600	IN	SOA	ciptapernika.id root.ciptapernika.id 2041430949 7200 3600 86400 3600

;; A Records
absensi.ciptapernika.id.	1	IN	A	103.143.170.181
ciptapernika.id.	1	IN	A	103.143.170.182
mail.ciptapernika.id.	1	IN	A	194.233.69.229
mailu.ciptapernika.id.	1	IN	A	194.233.69.229
mx.postal.ciptapernika.id.	1	IN	A	103.143.170.182
postal.ciptapernika.id.	1	IN	A	103.143.170.182
routes.postal.ciptapernika.id.	1	IN	A	103.143.170.182
rp.postal.ciptapernika.id.	1	IN	A	103.143.170.182
spf.postal.ciptapernika.id.	1	IN	A	103.143.170.182
track.postal.ciptapernika.id.	1	IN	A	103.143.170.182
webmin.ciptapernika.id.	1	IN	A	103.143.170.182

;; CNAME Records
psrp.ciptapernika.id.	1	IN	CNAME	rp.postal.ciptapernika.id.

;; MX Records
ciptapernika.id.	1	IN	MX	10 rp.postal.ciptapernika.id.
ciptapernika.id.	1	IN	MX	10 mx.postal.ciptapernika.id.
ciptapernika.id.	1	IN	MX	10 postal.ciptapernika.id.
mail.ciptapernika.id.	1	IN	MX	10 postal.ciptapernika.id.
routes.postal.ciptapernika.id.	1	IN	MX	10 postal.ciptapernika.id.

;; TXT Records
ciptapernika.id.	1	IN	TXT	"v=spf1 a mx include:spf.postal.ciptapernika.id ~all"
dkim._domainkey.mail.ciptapernika.id.	1	IN	TXT	"v=DKIM1; k=rsa; p=MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAxgM2SGQqycZrt6PIVd/dq/Zilast/boD3lF7DJb5m0a9nB245EcQsIJaYboARUjmNgxL9ichOAZSYau/mWeFp3m0ViCED4c1Q+frPr3XsiqxvO2MnkTljxbMVwQMSKBUsbIjfefgpxMDt3uVn5y+BPFKAj22xFqGHjyYAOdX2JArkWgkB3EpijaXfzOo;"
_dmarc.mail.ciptapernika.id.	1	IN	TXT	"v=DMARC1; p=reject; rua=mailto:admin@ciptapernika.id; ruf=mailto:admin@ciptapernika.id; adkim=s; aspf=s"
postal.ciptapernika.id.	1	IN	TXT	"v=spf1 a mx include:spf.postal.ciptapernika.id ~all"
postal-dyagps._domainkey.ciptapernika.id.	1	IN	TXT	"v=DKIM1; t=s; h=sha256; p=MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQC+AS0MNbWJUxzbtutJxAo2Hsmtkpck7mFjStPS1VC1UGADRSiXTQWTXtRql8QPnH5G+1iXZA0v5eufHcv3BcIFIcKMRfuFlAjNr45fu9kwGzTQ2a9QbGAxBO2ygilr67/7DS2Pt37ltgjJS2HJvejalD3sQJNOYatHD1r7l5i21wIDAQAB;"
postal-g5awdg._domainkey.ciptapernika.id.	1	IN	TXT	"v=DKIM1; t=s; h=sha256; p=MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQC8sCbuaF87Hy70lPUvzdfTw/XxYAU+q/y3SWCCasCqtUFxVxjeyXkfrxmmQA6bBnIRt/DqWBVy3a5qUo/7fS7xh+aK/WOd3jn4ymsKnjDa524mDqkekOPTOzIc5AaJwsE39pZhM/TI+FkrEESmYUIZUfGFbjab43DhsembHtL4AQIDAQAB;"
postal-gua04l._domainkey.postal.ciptapernika.id.	1	IN	TXT	"v=DKIM1; t=s; h=sha256; p=MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQCoIVlehqsc8VH/m266N5VSReJZGbUYndmuILXOtTYDHeZDYCo1++QDnFmuzIzYx+lg3bCNM7p0eKtabMZpe9fQeR/pTpalBtGD++u2GLIaI+ztFzNRgRrdO/DlDYaw2Xrv5PJ59Ser/D+gyIn/A/j+aStlgcBg6Ip7XN35eVlE0QIDAQAB;"
postal-ua0o0v._domainkey.ciptapernika.id.	1	IN	TXT	"v=DKIM1; t=s; h=sha256; p=MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQCj4raXlfFF0QN/dG4Z72BJUSbDYbDE2IPUwLpzGhL25ll6TlmQfufMQDpTv4UrzOVa+gBFS8MHoM1pmsX7KZYOEDxyAiO1Et61pNdu6mtYjPD+tMyA9weYlRWJFkhvOno8111VLmWn+6Qpre7UILMJkx+0re63pxFcdHll0lD04wIDAQAB;"
postal-vvqtqr._domainkey.ciptapernika.id.	1	IN	TXT	"v=DKIM1; t=s; h=sha256; p=MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQDZijtjd/Ap2gvnY4OR302jNl1u2lCrzxrndCsu98koyXPgDW6LoveSwHfRW0ssUHbqTyoPRyYa9PjN2qfp22ut2QiPunDBdqp+qH9QgYNjgh7SaTTfkSKhohRrRELivIE9SsJy23tNLqYXp1b1PyY6MC5Gs/Sn6ZFPH0j2+kwC6QIDAQAB;"
postal-wnmrnj._domainkey.ciptapernika.id.	1	IN	TXT	"v=DKIM1; t=s; h=sha256; p=MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQDnww4wbwehEbB5DF0FzZ4LUjJNFqzkoMgzUdf1xE8uN1rZ8GCc2W3Lt18J81LA4MGU+74Gr5FKg4qCoec561z+UYPSqkwhpIReWHCx/s0C/DScKXwdtzl7AYEVLP3sDAARIKq2rD63ttE6l9cbvYD/xiKcIu1pW4SybSBIyeGtOQIDAQAB;"
postal-zz3yan._domainkey.ciptapernika.id.	1	IN	TXT	"v=DKIM1; t=s; h=sha256; p=MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQC1SUAlnf3NcXelCW3aoBY0LZhLFRpX/BKf3s9nHZh4TcYL2YsM6ECSCnWuPNNUtQdtPehJ4jhTkSZHI7SgO0S2sL4guGbFJs4rjYcN6BVooJUg5UNExvGNJq7ZyzkqwpEgNCE2RgGlfWQ4W6g6gzYFzAWx/pqn8umyqOaiZCSZZwIDAQAB;"
rp.postal.ciptapernika.id.	1	IN	TXT	"v=spf1 a mx include:spf.postal.ciptapernika.id ~all"
spf.postal.ciptapernika.id.	1	IN	TXT	"v=spf1 ip4:103.143.170.182 ~all"
